### PR TITLE
bme280 driver in Lua+C

### LINKED ISSUE #3126: Rewrite bme280 driver in Lua+C
In an effort to decrease the project's technical debt, I've proposed (in #3124) that the C-only `bme280` be cut into a Lua layer that drives communication and a C module that does any requisite low-level fiddling.  This will be similar to #2819 and may, incidentally, address #2527 correctly as well.  (As you can see, between that and #2241, this module has a history of being sufficiently troublesome that issues have been worked around rather than fixed.)  This rewrite will *also* eliminate yet another source of static allocation and artificial limitation in our driver layer.  All in all, deeply worth it.

I would like it very much if the result of this rewrite included a test driver, as well.  Ideally `bme280` becomes an example of how all drivers should look.  No pressure (ha!). ;)

OK, so, what does this entail, concretely?

Looking at the existing `bme280.c`, we almost immediately see that there's a bunch of goo associated with performing `i2c` transactions: `bme280_i2c_addr`, ` r8u_n`, `r8u`, and `w8u`.  These can all be directly transcoded to Lua.  The results from the `i2c` package will be Lua strings.

The `bme280_data` structure holds per-device state obtained during `setup`, which means we probably want to hold it instead in a Lua userdata object that represents the sensor.  Reading `setup`, it looks like this takes 3 or 4 I2C transactions to set up, including getting the chip ID, so I think our sensor object constructor should take 2 or 3 Lua strings (`DIG_T`, `DIG_P`, and `DIG_H`) to build itself.  Since it appears that the static values `bme280_t_fine`, `bme280_h`, and `bme280_hc` may outlive a particular measurement, they should perhaps join the `bme280_data` in the per-device userdata object.

Armed with the sensor parameters, we can look at interacting with the device: `bme280_lua_startreadout` does some i2c transactions, easily done instead in Lua, and schedules a timer, also easily done in Lua.  (Can the device not raise IRQs?)

Once it's time to read data, `bme280_lua_read` does some more i2c transactions and a lot of math, including storing "fine" constants between procedure calls.  To continue the theme, the transactions should be done in Lua, and passed to a C routine that takes the sensor userdata object and does the requisite math.  For my $0.02, the results should just land in a Lua table.

The abbreviated `bme280_lua_temp`, `bme280_lua_baro`, and `bme280_lua_humi` functions are readily achieved by having the Lua replacement for `bme280_lua_read` perform fewer transactions and call into the C function with fewer string arguments.  That function should return a table with fewer keys as a result.

For the purposes of PRing, I'd suggest placing code in

* `app/modules/bme280_math.c` for the C layer
* `lua_modules/bme280/bme280.lua` for the Lua top-level and main driver
* `lua_modules/bme280/bme280_test.lua` for the test script (ideally in mispec, I think?  See #2145).

### PR DESCRIPTION
**Pull request by vsky279**
Fixes #3126, #3124, #2241, 

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/*`.

Rewritten BME280 driver based on the discussion in #3126.
I intend to add also SPI support (#2527) which should be also relatively small modification of the Lua driver.

There is a conflict when original BME280 module is included in the build. `require('bme280')` references to the original C module and not to the new Lua driver. This can be a bit confusing. The natural solution would be to that this driver replaces the original C driver.

I assume this should be probably merged only when #3128 is implemented.

Please review the proposed commit.

### CODE REVIEW

**Comment by vsky279** [comment]
I don't know how to make Travis CI happy when it complains that
```lua_modules/bme280/bme280.lua:22:3: accessing undefined variable bme280_math```
`bme280_math` is the new module so it should be known to the builder.

**Comment by galjonsfigur** [comment]
@vsky279 You have to add entities in `luacheck` config file that can be found in `tools/luacheck_config.lua`. You can write it yourself or use luacheck config helper(also in `tools` folder) script that will generate this for you - little note on how to use it can be found [here](https://github.com/nodemcu/nodemcu-firmware/blob/master/tools/luacheck_config_helper.lua#L4-L6).

I thought that I've written documentation on how to add new entities in luacheck config when adding new C module but turns out I was wrong - I could add this in documentation or in project wiki.

**Comment by vsky279** [comment]
@galjonsfigur Thanks. I did manual update and now it's green.

**Comment by vsky279** [comment]
#### Lua 5.3 compatibility issue - `lua_pushfstring` works differently for non-printable characters
I was using `lua_pushfstring` with `%c` formating string to return string array of configuration parameters (3 bytes). Since Lua 5.3 the string pushed to stack for e.g. value 240 is `<\240>`. Different approach using just `lua_pushstring` is needed.

I am not sure if this cannot cause some issues in other modules.

https://github.com/nodemcu/nodemcu-firmware/blob/8d091c476edf6ae2977a5f2a74bf5824d07d6183/app/lua53/lobject.c#L426

File: mkdocs.yml [hunk 1]
```diff
@@ -48,6 +48,10 @@ pages:
 - Lua Modules:
     - 'Lua modules directory': 'lua-modules/README.md'
     - 'bh1750': 'lua-modules/bh1750.md'
+<<<<<<< HEAD
+=======
+    - 'bme280':  'lua-modules/bme280.md'
+>>>>>>> Initial commit
```
Comment (nwf) [inline mkdocs.yml:54]:
```suggestion
    - 'bme280':  'lua-modules/bme280.md'
```

**Pull request merged**